index.js

```diff
@@ -1,60 +1,72 @@
-'use strict';
-
-var React = require('react'),
-    invariant = require('react/lib/invariant'),
-    shallowEqual = require('react/lib/shallowEqual');
-
-function createSideEffect(onChange, mixin) {
-  invariant(
-    typeof onChange === 'function',
-    'onChange(propsList) is a required argument.'
-  );
-
-  var mountedInstances = [];
-
-  function emitChange() {
-    onChange(mountedInstances.map(function (instance) {
-      return instance.props;
-    }));
-  }
-
-  return React.createClass({
-    mixins: [mixin],
-
-    statics: {
-      dispose: function () {
-        mountedInstances = [];
-        emitChange();
-      }
-    },
-
-    shouldComponentUpdate: function (nextProps) {
-      return !shallowEqual(nextProps, this.props);
-    },
-
-    componentWillMount: function () {
-      mountedInstances.push(this);
-      emitChange();
-    },
-
-    componentDidUpdate: function () {
-      emitChange();
-    },
-
-    componentWillUnmount: function () {
-      var index = mountedInstances.indexOf(this);
-      mountedInstances.splice(index, 1);
-      emitChange();
-    },
-
-    render: function () {
-      if (this.props.children) {
-        return React.Children.only(this.props.children);
-      } else {
-        return null;
-      }
+import React from "react";
+import invariant from "react/lib/invariant";
+import shallowEqual from "react/lib/shallowEqual";
+
+const RESERVED_PROPS = {
+    arguments: true,
+    caller: true,
+    key: true,
+    length: true,
+    name: true,
+    prototype: true,
+    ref: true,
+    type: true
+};
+
+export default (Component) => {
+    invariant(
+        typeof Object.is(Component.handleChange, "function"),
+        "handleChange(propsList) is not a function."
+    );
+
+    const mountedInstances = new Set();
+    const emitChange = () => {
+        Component.handleChange([...mountedInstances].map(instance => instance.props));
+    };
+
+    class CreateSideEffect extends React.Component {
+        static displayName = "CreateSideEffect"
+
+        componentWillMount() {
+            mountedInstances.add(this);
+            emitChange();
+        }
+
+        shouldComponentUpdate(nextProps) {
+            return !shallowEqual(nextProps, this.props);
+        }
+
+        componentDidUpdate() {
+            emitChange();
+        }
+
+        componentWillUnmount() {
+            if (mountedInstances.has(this)) {
+                mountedInstances.delete(this);
+            }
+
+            emitChange();
+        }
+
+        static dispose() {
+            mountedInstances.clear();
+            emitChange();
+        }
+
+        render() {
+            return (
+                <Component {...this.props} />
+            );
+        }
     }
-  });
-}
 
-module.exports = createSideEffect;
\ No newline at end of file
+    Object.getOwnPropertyNames(Component)
```

Comment:
This copies the static methods from `Component` so they are still available in the wrapped component.